### DISCUSSION TITLE
Provide event handlers with all the arguments they request

### DISCUSSION THREAD

**Discussion by drzraf** [miguelgrinberg/python-socketio#1362] [Ideas]
**Is your feature request related to a problem? Please describe.**

Arguments set passed to handlers is very opinionated and overly restricted.
I can't access `namespace`, `sid`, `event`'s name whenever I want.
It'd be easy to pass keyword arguments and let handlers do whatever they want with them. 

```python
def log(self, event, data):
  # Do something with {event} & {data}

# Made a loop to avoid cumbersome repetitions due to wrapping
# my handler... for the sole purpose of providing `evt`
# but it doesn't make the code clearer
for evt in ['foo', 'bar']:
    sio.on(evt, handler=lambda data: log(evt, data))
    
```

**Describe the solution you'd like**
```python

def log(self, event, data, sid, namespace):
  # Do something with {*}
  # or maybe **kwarg and let me choose for it
  # or any solution that allows me to access *all* properties of this message

sio.on('foo', namespace='/', handler=log)
sio.on('bar', handler=log)
```

**Comment by miguelgrinberg** [comment]
Have you seen the catch-all event handlers in the documentation? If you create a catch-all namespace and event handler you'll get what you want, I think.

https://python-socketio.readthedocs.io/en/stable/client.html#catch-all-event-and-namespace-handlers

**Reply by drzraf** [reply]
I want to keep the catch-all for other purposes.

I'd prefer to run my functions for specific events and still retrieve all the context (sid/namespace/...).
(I also have a case where I needed the `sid` (to later disconnect) which isn't possible provided to named event handler)

**Reply by miguelgrinberg** [reply]
Sounds good. You can add any arguments you want on your handler functions using `functools.partial`. The sid is passed on to all event handlers in the server. In the client the sid is the same for all events. You can call the [sio.get_sid()](https://python-socketio.readthedocs.io/en/stable/api.html#socketio.Client.get_sid) method to get it.
